make dev-env is broken

- `make build image` and `make dev-env` broken https://kubernetes.github.io/ingress-nginx/developer-guide/getting-started/#local-build
-
% make clean && make dev-env                                                                                                                                                                                                                       
[dev-env] building image                                                                                                                                                                                                                           
..printing env & other vars to stdout                                                                                                                                                                                                              
HOSTNAME=mypc                                                                                                                                                                                                                                      
Darwin mypc 21.5.0 Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:22 PDT 2022; root:xnu-8020.121.3~4/RELEASE_X86_64 x86_64                                                                                                                         
TAG=1.0.0-dev                                                                                                                                                                                                                                      
TERM_PROGRAM=tmux                                                                                                                                                                                                                                  
COMMIT_SHA=git-60b714336                                                                                                                                                                                                                           
SHELL=/bin/zsh                                                                                                                                                                                                                                     
TERM=screen-256color                                                                                                                                                                                                                               
MAKEFLAGS=s                                                                                                                                                                                                                                        
FOCUS=.*                                                                                                                                                                                                                                           
TMPDIR=/var/folders/gk/tc355p1112z93mwtq88j9wnm0000gq/T/                                                                                                                                                                                           
TERM_PROGRAM_VERSION=3.3a                                                                                                                                                                                                                          
TERM_SESSION_ID=B0A9FC47-7E70-403E-8328-51A715C13C2B                                                                                                                                                                                               
PNPM_HOME=/Users/m/Library/pnpm                                                                                                                                                                                                                    
USER=m                                                                                                                                                                                                                                             
CURDIR=/Users/m/Documents/github/longwuyuan/ingress-nginx                                                                                                                                                                                          
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.iUXG0Fb4Hf/Listeners                                                                                                                                                                                  
__CF_USER_TEXT_ENCODING=0x1F7:0x0:0x2                                                                                                                                                                                                              
PLATFORMS=amd64 arm arm64 s390x                                                                                                                                                                                                                    
MAKELEVEL=2                                                                                                                                                                                                                                        
MFLAGS=-s                                                                                                                                                                                                                                          
TMUX=/private/tmp/tmux-503/default,1419,0   
PATH=/Users/m/Library/pnpm:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/m/bin:/Users/m/.local/bin:/Users/m/go/bin:/usr/local/sbin:/Users/m/.npm-global
__CFBundleIdentifier=com.apple.Terminal
PWD=/Users/m/Documents/github/longwuyuan/ingress-nginx
HOST_ARCH=amd64
E2E_CHECK_LEAKS=
MAC_OS=Darwin
GOARCH=amd64
EDITOR=nvim
PKG=k8s.io/ingress-nginx
TMUX_PANE=%1
XPC_FLAGS=0x0
BUILD_ID="UNSET"
E2E_NODES=7
XPC_SERVICE_NAME=0
HOME=/Users/m
SHLVL=6
BUILDX_PLATFORMS=linux/amd64,linux/arm,linux/arm64,linux/s390x
BASE_IMAGE=registry.k8s.io/ingress-nginx/nginx:0ff500c23f34e939305de709cb6d47da34b66611@sha256:15f91034a03550dfab6ec50a7be4abbb683d087e234ad7fef5adedef54e46a5a
LOGNAME=m
LC_CTYPE=UTF-8
MAKEFILE_LIST= Makefile
ARCH=amd64
REPO_INFO=git@github.com:longwuyuan/ingress-nginx.git
MAC_DOCKER_FLAGS=
REGISTRY=gcr.io/k8s-staging-ingress-nginx
_=/usr/bin/env
DIND_ENABLED=
done..printing env & other vars to stdout
Reached DIND check ELSE block, inside run-in-docker.sh
+ GO_BUILD_CMD='go build -v'
+ set -o errexit
+ set -o nounset
+ set -o pipefail
+ declare -a mandatory
+ mandatory=(PKG ARCH COMMIT_SHA REPO_INFO TAG)
+ for var in "${mandatory[@]}"
+ [[ -z k8s.io/ingress-nginx ]]
+ for var in "${mandatory[@]}"
+ [[ -z amd64 ]]
+ for var in "${mandatory[@]}"
+ [[ -z git-60b714336 ]]
+ for var in "${mandatory[@]}"
+ [[ -z git@github.com:longwuyuan/ingress-nginx.git ]]
+ for var in "${mandatory[@]}"
+ [[ -z 1.0.0-dev ]]
+ export CGO_ENABLED=0
+ CGO_ENABLED=0
+ export GOARCH=amd64
+ GOARCH=amd64
+ TARGETS_DIR=rootfs/bin/amd64
+ echo 'Building targets for amd64, generated targets in rootfs/bin/amd64 directory.'
Building targets for amd64, generated targets in rootfs/bin/amd64 directory.
+ echo 'Building k8s.io/ingress-nginx/cmd/nginx'
Building k8s.io/ingress-nginx/cmd/nginx
+ echo f61981049b33
f61981049b33
+ go build -v -trimpath '-ldflags=-buildid= -w -s     -X k8s.io/ingress-nginx/version.RELEASE=1.0.0-dev     -X k8s.io/ingress-nginx/version.COMMIT=git-60b714336     -X k8s.io/ingress-nginx/version.REPO=git@github.com:longwuyuan/ingress-nginx.git' -o rootfs/bin/amd64/nginx-ingress-controller k8s.io/ingress-nginx/cmd/nginx
# cd /go/src/k8s.io/ingress-nginx; git status --porcelain
fatal: unsafe repository ('/go/src/k8s.io/ingress-nginx' is owned by someone else)
To add an exception for this directory, call:

        git config --global --add safe.directory /go/src/k8s.io/ingress-nginx
error obtaining VCS status: exit status 128
        Use -buildvcs=false to disable VCS stamping.
make[1]: *** [build] Error 1
make: *** [dev-env] Error 2

Yes, because this is the error ;
```
# cd /go/src/k8s.io/ingress-nginx; git status --porcelain
fatal: unsafe repository ('/go/src/k8s.io/ingress-nginx' is owned by someone else)
To add an exception for this directory, call:

        git config --global --add safe.directory /go/src/k8s.io/ingress-nginx
error obtaining VCS status: exit status 128
        Use -buildvcs=false to disable VCS stamping.
make[1]: *** [build] Error 1
make: *** [dev-env] Error 2
```

And I just now added that to build.sh in my branch + blanked the variable $MAC_OS and I was able to build successfully

I have limited understanding of this.

But my experience is that I had a perfectly working environment to run `make dev-env` and `make build image`. Now that is broken due to 2 reasons mentioned below.

(1) Since git version 2.35.2 there is a fix in git that requires use of safe.directory. My guess is that my environment was on older versions of git earlier. And now got got updated and the error is what I have described above already ;
```
# cd /go/src/k8s.io/ingress-nginx; git status --porcelain
fatal: unsafe repository ('/go/src/k8s.io/ingress-nginx' is owned by someone else)
To add an exception for this directory, call:

        git config --global --add safe.directory /go/src/k8s.io/ingress-nginx
error obtaining VCS status: exit status 128
        Use -buildvcs=false to disable VCS stamping.
make[1]: *** [build] Error 1
make: *** [dev-env] Error 2
```
I am convinced about this because I simulate the CI pipeline run. I do launch `make dev-env` from my laptop's os shell, but ultimately the run-in-docker.sh etc causes the build to occur in a container in DIND. There its not me as user but whatever  (see screenshot) below
<img width="1714" alt="image" src="https://user-images.githubusercontent.com/5085914/178173834-cb4ec6a8-7e15-49b2-8a0a-f56e16d28991.png">

As you can see, I have done the `safe.directory` thing that git recommends. If I don't do this in the script then build fails. There is no use of setting safe.directory on my laptop os git workspace because the build is actually happening in a container in DIND.

(2) The variable `MAC_OS` in the Makefile is set to value `Darwin` on my laptop during this build. This causes another variable `MAC_DOCKER_FLAGS`  to be set to a value of `--load`. This is a new change. If I let this be, then build fails for me with error message as screenshot below ;
<img width="351" alt="image" src="https://user-images.githubusercontent.com/5085914/178174450-2b817c87-1e3a-4dea-b679-1c5079d25d02.png">

The `--load` flag is invalid for the build running in the container in DIND because that is linux and not `MAC_OS=Dawrwin`.  So I set the value to space `" "` in the Makefile, where it was being set to `--load`. And voila my build was successful.

So far I gather that `build.sh` or other scripts can be run manually on laptop. But changes for that should not break the `make dev-env` or `make build image`. I am not sure what is appropriate here, and I have seen Ricardo mention that the var needs to be initialised .  But hopefully we don't break `make dev-env` or `make build image` .

--- a/Makefile
+++ b/Makefile
@@ -54,8 +54,9 @@ ifneq ($(PLATFORM),)
 endif
 
 MAC_OS = $(shell uname -s)
+
 ifeq ($(MAC_OS), Darwin)
-	MAC_DOCKER_FLAGS="--load"
+	MAC_DOCKER_FLAGS=
 else
 	MAC_DOCKER_FLAGS=
 endif

--- a/build/build.sh
+++ b/build/build.sh
@@ -49,6 +49,7 @@ echo "Building targets for ${ARCH}, generated targets in ${TARGETS_DIR} director
 
 echo "Building ${PKG}/cmd/nginx"
 
+git config --add safe.directory /go/src/k8s.io/ingress-nginx
 ${GO_BUILD_CMD} \
   -trimpath -ldflags="-buildid= -w -s \
     -X ${PKG}/version.RELEASE=${TAG} \

--- a/build/run-in-docker.sh
+++ b/build/run-in-docker.sh
@@ -65,12 +65,14 @@ fi
 
 USER=${USER:-nobody}
 
+MAC_OS="`uname -s`"
 MAC_OS="${MAC_OS:-}"
 if [[ ${MAC_OS} == "Darwin" ]]; then
 	MAC_DOCKER_FLAGS=""
 else
 	MAC_DOCKER_FLAGS="-u $(id -u ${USER}):$(id -g ${USER})" #idk why mac/git fails on the gobuild if these are presented to dockerrun.sh script
 fi
+echo "MAC_OS = ${MAC_OS}, MAC_OS_FLAGS = ${MAC_DOCKER_FLAGS}"
 
 echo "..printing env & other vars to stdout"
 echo "HOSTNAME=`hostname`"

